### PR TITLE
Add Index Bounds to Queries

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds functionality to `Target` to compute boundaries for Index scans. 

See how this is used in the prototype: https://github.com/firebase/firebase-android-sdk/pull/2654/files#diff-979270768a3e2798683bf7702f4e84ade098aa967d60d90c9c741d41fc64754dR191 Note that the prototype uses multiple methods instead of the API in this PR.

### CODE REVIEW

File: firebase-firestore/src/test/java/com/google/firebase/firestore/core/TargetTest.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+import java.util.Collections;
+import java.util.List;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(manifest = Config.NONE)
+public class TargetTest {
```
Comment (wu-hui) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/core/TargetTest.java:39]: I think we support partial index match. Should we add some tests for that?

**Review by wu-hui — CHANGES_REQUESTED** [review]
Sorry for the delay!

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java [hunk 2]
```diff
@@ -107,8 +111,146 @@ public boolean hasLimit() {
     return endAt;
   }
 
+  /**
+   * Returns a lower bound of field values that can be used as a starting point to scan the index
+   * defined by `fieldIndex`.
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java [hunk 3]
```diff
@@ -107,8 +111,146 @@ public boolean hasLimit() {
...
+   *
+   * <p>Unlike {@link #getUpperBound}, lower bounds always exist as the SDK can use {@code null} as
+   * a starting point for missing boundary values.
+   *
+   * @param fieldIndex The index definition to generate a Bound for
+   * @return A bound that can be used for index scanning
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java [hunk 4]
```diff
@@ -107,8 +111,146 @@ public boolean hasLimit() {
...
+      // If there is a startAt bound, compare the values against the existing boundary to see
+      // if we can narrow the scope.
+      if (startAt != null) {
+        for (int i = 0; i < orderBys.size(); ++i) {
+          OrderBy orderBy = this.orderBys.get(i);
+          if (orderBy.getField().equals(segment.getFieldPath())) {
+            Value cursorValue = startAt.getPosition().get(i);
+            if (Values.compare(lowestValue, cursorValue) < 0) {
+              lowestValue = cursorValue;
+              before = before && startAt.isBefore();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java:162]: `before` is shared by all cursor values (it applies to `values` and not just `lowestValue`.  This code might be a bit misleading though as it suggests that a new value can be found for the specific field that we are processing. I don't think that is true and we can exit early here.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java [hunk 5]
```diff
@@ -107,8 +111,146 @@ public boolean hasLimit() {
...
+      }
+
+      // If there is an endAt bound, compare the values against the existing boundary to see
+      // if we can narrow the scope.
+      if (endAt != null) {
+        for (int i = 0; i < orderBys.size(); ++i) {
+          OrderBy orderBy = this.orderBys.get(i);
+          if (orderBy.getField().equals(segment.getFieldPath())) {
+            Value cursorValue = endAt.getPosition().get(i);
+            if (largestValue == null) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java]: Updated

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java [hunk 6]
```diff
@@ -107,8 +111,146 @@ public boolean hasLimit() {
...
+              largestValue = cursorValue;
+            } else if (Values.compare(largestValue, cursorValue) > 0) {
+              largestValue = cursorValue;
+            }
+            before = endAt.isBefore();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java]: I think you are right. This is a bit hard to get my head around. I will add a test.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/core/TargetTest.java (hunk 1 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/core/TargetTest.java:39]: Added a test.

**Review by schmidt-sebastian — COMMENTED** [review]
Thanks for the review!

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java:162]: I see, can we add a comment here then?

**Review by wu-hui — APPROVED** [review]

**Pull request merged**